dojo/tools/dependency_track/parser.py

```diff
@@ -0,0 +1,202 @@
+                    "uuid": "701a3953-666b-4b7a-96ca-e1e6a3e1def3",
+                    "source": "NPM",
+                    "vulnId": "48",
+                    "title": "Regular Expression Denial of Service",
+                    "subtitle": "uglify-js",
+                    "severity": "LOW",
+                    "severityRank": 3,
+                    "cweId": 400,
+                    "cweName": "Uncontrolled Resource Consumption ('Resource Exhaustion')",
+                    "description": "Versions of `uglify-js` prior to...",
+                    "recommendation": "Update to version 2.6.0 or later."
+                },
+                "analysis": {
+                    "isSuppressed": false
+                },
+                "matrix": "ca4f2da9-0fad-4a13-92d7-f627f3168a56:979f87f5-eaf5-4095-9d38-cde17bf9228e:701a3953-666b-4b7a-96ca-e1e6a3e1def3"
+            }]
+    }
+    """
+
+    def _convert_dependency_track_severity_to_dojo_severity(self, dependency_track_severity):
+        """
+        Converts a Dependency Track severity to a DefectDojo severity.
+        :param dependency_track_severity: The severity from Dependency Track
+        :return: A DefectDojo severity if a mapping can be found; otherwise a null value is returned
+        """
+        severity = dependency_track_severity.lower()
+        if severity == "critical":
+            return "Critical"
+        elif severity == "high":
+            return "High"
+        elif severity == "medium":
+            return "Medium"
+        elif severity == "low":
+            return "Low"
+        elif severity.startswith("info"):
+            return "Informational"
+        else:
+            return None
+
+    def _convert_dependency_track_finding_to_dojo_finding(self, dependency_track_finding, test):
+        """
+        Converts a Dependency Track finding to a DefectDojo finding
+
+        :param dependency_track_finding: A dictionary representing a single finding from a Dependency Track Finding Packaging Format (FPF) export
+        :param test: The test that the DefectDojo finding should be associated to
+        :return: A DefectDojo Finding model
+        """
+        # Build the title of the Dojo finding
+        if 'vulnerability' not in dependency_track_finding:
+            raise Exception("Missing 'vulnerability' node from finding!")
+        if 'vulnId' not in dependency_track_finding['vulnerability']:
+            raise Exception("Missing 'vulnId' node from vulnerability!")
+        vuln_id = dependency_track_finding['vulnerability']['vulnId']
+        if 'source' not in dependency_track_finding['vulnerability']:
+            raise Exception("Missing 'source' node from vulnerability!")
+        source = dependency_track_finding['vulnerability']['source']
+        if 'component' not in dependency_track_finding:
+            raise Exception("Missing 'component' node from finding!")
+        if 'purl' not in dependency_track_finding['component']:
```

Comment:
Not all findings will have a purl. If a component is of type (operating system or hardware device), these components will not have a purl. Software components without a purl are also likely and may involve unmanaged dependencies, dependencies that were manually added, file components.

Reply:
Hmm interesting. Good to know! Any thoughts on how we should build the title/description of a finding in DefectDojo given the data from Dependency Track. Are there attributes of the component that are guaranteed to be there (e.g. name, version, etc.)?

Reply:
Component name is the only field that is required. You can use the UUID of the vulnerability to make an additional query to retrieve the full description, recommendation, credits, etc. 

The vulnerability in a finding will always include uuid, source, and vulnid, and may optionally include a title and subtitle.

Reply:
The way DefectDojo seems to work is you upload a file (either manually in the UI or via the API) that contains the scan results. This would need to be self contained and parsed for all of the needed info.

Do you have some example FPF JSON payloads you can share that show some of these things:

Component:
- No purl
- Only component name (no uid/version)

Vulnerability:
- has a title and/or subtitle

Reply:
Dependency Track seems to indicate that the `version` of a component is required (as indicated by the asterisk *).

![image](https://user-images.githubusercontent.com/51127567/65257575-69d28300-dacf-11e9-9842-6421d2729f77.png)

As such why would the version not come back in the FPF for a Component?

Reply:
Components will also have a uuid as well. That's a requirement. But version is optional.

Also note that the findings API is just one way to get the results. The vulnerability api may also be used, but you won't get the component or analysis.

FYI, ThreadFix uses the vulnerability API, Kenna Security uses the Findings API that is then converted to KDI prior to publish.

Reply:
Wait hold. On. The vulnerability description and recommendation WILL be populated in the findings results if those values are non-null.

Reply:
if the case of CVE, this means that a vulnerabiliy will generally have a description. In the case of NPM Audit findings, the vuln will typically have a description and recommendation.

Reply:
The 'version' is not a requirement in the data model. There might be some business logic enforcement to ensure version is there, but to be safe, I would account for instances where version is null

Reply:
Okay I made a few updates to accommodate for these optional fields from Dependency Track. Let me know what you think. Diff is here: https://github.com/DefectDojo/django-DefectDojo/pull/1561/commits/ff1fb01da3d86f3f4840e30c814a4b8d1999c98f